quic/s2n-quic-core/src/recovery/bandwidth/estimator/tests.rs

```diff
@@ -0,0 +1,328 @@
+    );
+    assert_eq!(new_packet.time_sent, bw_estimator.state.first_sent_time);
+    assert_eq!(t1 - t0, bw_estimator.rate_sample.interval);
+
+    // Ack an older packet
+    let mut old_packet = new_packet;
+    old_packet.bandwidth_state.delivered_bytes -= 1;
+    old_packet.time_sent -= Duration::from_secs(1);
+
+    bw_estimator.on_packet_ack(&packet, t1);
+
+    assert_eq!(
+        bw_estimator.state.delivered_bytes,
+        (packet.sent_bytes + new_packet.sent_bytes + old_packet.sent_bytes) as u64
+    );
+    assert_eq!(
+        bw_estimator.rate_sample.newly_acked_bytes,
+        (packet.sent_bytes + new_packet.sent_bytes + old_packet.sent_bytes) as u64
+    );
+
+    // Rate sample is not updated since this packet is older than the current sample
+    assert_eq!(
+        new_packet.bandwidth_state.delivered_bytes,
+        bw_estimator.rate_sample.prior_delivered_bytes
+    );
+    assert_eq!(
+        new_packet.bandwidth_state.lost_bytes,
+        bw_estimator.rate_sample.prior_lost_bytes
+    );
+    assert_eq!(
+        new_packet.bandwidth_state.is_app_limited,
+        bw_estimator.rate_sample.is_app_limited
+    );
+    assert_eq!(
+        new_packet.bytes_in_flight,
+        bw_estimator.rate_sample.bytes_in_flight
+    );
+    assert_eq!(new_packet.time_sent, bw_estimator.state.first_sent_time);
+    assert_eq!(t1 - t0, bw_estimator.rate_sample.interval);
+}
+
+//= https://tools.ietf.org/id/draft-cheng-iccrg-delivery-rate-estimation-02#2.2.4
+//= type=test
+//# Since it is physically impossible to have data delivered faster than it is sent
+//# in a sustained fashion, when the estimator notices that the ack_rate for a flight
+//# is faster than the send rate for the flight, it filters out the implausible ack_rate
+//# by capping the delivery rate sample to be no higher than the send rate.
+#[test]
+fn on_packet_ack_implausible_ack_rate() {
+    let t0 = NoopClock.get_time();
+    let mut bw_estimator = Estimator::new(t0);
+    bw_estimator.state.delivered_time = t0 + Duration::from_secs(4);
+
+    let packet = SentPacketInfo::new(
+        true,
+        1500,
+        t0 + Duration::from_secs(5),
+        AckElicitation::Eliciting,
+        path::Id::test_id(),
+        ExplicitCongestionNotification::NotEct,
+        bw_estimator.state(),
+        500,
+    );
+
+    let now = packet.time_sent + Duration::from_secs(1);
+    bw_estimator.on_packet_ack(&packet, now);
+
+    let send_elapsed = packet.time_sent - packet.bandwidth_state.first_sent_time;
+    let ack_elapsed = now - packet.bandwidth_state.delivered_time;
+
+    // The amount of time to send the packets was greater than the time to
+    // acknowledge them, indicating the ack rate would be implausible.
+    assert!(send_elapsed > ack_elapsed);
+
+    // The rate sample interval is based on the send_elapsed time since the
+    // ack_elapsed time was implausible
+    assert_eq!(send_elapsed, bw_estimator.rate_sample.interval);
```

Comment:
what would cause this to happen? is it worth having an event when this occurs?

Reply:
From [the RFC](https://www.ietf.org/archive/id/draft-cheng-iccrg-delivery-rate-estimation-02.html#name-compression-and-aggregation): 

> For computing the delivery_rate, the sender prefers ack_rate, the rate at which packets were acknowledged, since this usually the most reliable metric. However, this approach of directly using "ack_rate" faces a challenge when used with paths featuring aggregation, compression, or ACK decimation, which are prevalent [A15]. In such cases, ACK arrivals can temporarily make it appear as if data packets were delivered much faster than the bottleneck rate.

I don't think it would be too useful to have an event for this, but we can always add one later